pkg/buildpipeline/pipeline.go

```diff
@@ -155,6 +155,61 @@ func (s *Stage) TaskName() string {
 	return strings.ToLower(strings.NewReplacer(" ", "-").Replace(s.Name))
 }
 
+// Task names need to be RFC 1035/1123 compliant DNS labels, so we mangle the
+// input to make it compliant. Results should match the following regex and be
+// no more than 63 characters long:
+//     [a-z]([-a-z0-9]*[a-z0-9])?
+// cf. https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+// Suffix is assumed to be alphanumeric and non-empty.
+func createTaskName(pipelineIdentifier string, stageName string, buildIdentifier string, suffix string) string {
+	const MAX_LABEL_LENGTH = 63
+	MAX_BODY_LENGTH := MAX_LABEL_LENGTH - len(suffix) - 1 // Add an extra hyphen before the suffix
+
+	// Having `stage` and `build` literally in the output means the body is
+	// guaranteed to be non-empty. Would it be better to fail if the output is
+	// `stage-build-$suffix` because everything else is invalid?
+	body := fmt.Sprintf("%s-stage-%s-build-%s", pipelineIdentifier, stageName, buildIdentifier)
+
+	var sb strings.Builder
+	bufferedHyphen := false // Used to make sure we don't output consecutive hyphens.
+	for _, codepoint := range body {
```

Comment:
I gotta wonder whether there might be something like this in some library we already pull in...but hey.

Reply:
Yeah very possible. There is https://github.com/kubernetes/kubernetes/blob/0b8566f3887a2d13baba623d88aeb64f6e637f46/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L162 to just say "is this valid?", but I didn't see any mangling functions.

Reply:
Turns out JX already a function that does most of what we want, but it doesn't make capital letters lowercase or enforce a maximum length: https://github.com/jenkins-x/jx/blob/43f6dae88b40a139572758149d92df6d366796ab/pkg/kube/names.go#L18.